sympy/sets/ordinals.py

```diff
@@ -0,0 +1,257 @@
+from sympy.core import Basic, Integer
+from sympy.core.compatibility import with_metaclass
+from sympy.core.singleton import Singleton, S
+import operator
+
+class OmegaPower(Basic):
+    """
+    Represents ordinal exponential and multiplication terms one of the
+    building blocks of the Ordinal class.
+    In OmegaPower(a, b) a represents exponent and b represents multiplicity.
+    """
+    def __new__(cls, a, b):
+        if isinstance(b, int):
+            b = Integer(b)
+        if not isinstance(b, Integer) or b <= 0:
+            raise TypeError("multiplicity must be a positive integer")
+
+        if not isinstance(a, Ordinal):
+            a = Ordinal.convert(a)
+
+        return Basic.__new__(cls, a, b)
+
+    @property
+    def exp(self):
+        return self.args[0]
+
+    @property
+    def mult(self):
+        return self.args[1]
+
+    def _compare_term(self, other, op):
+        if self.exp == other.exp:
+            return op(self.mult, other.mult)
+        else:
+            return op(self.exp, other.exp)
+
+    def __eq__(self, other):
+        if not isinstance(other, OmegaPower):
+            try:
+                other = OmegaPower(0, other)
+            except TypeError:
+                return NotImplemented
+        return hash(self) == hash(other)
+
+    def __lt__(self, other):
+        if not isinstance(other, OmegaPower):
+            try:
+                other = OmegaPower(0, other)
+            except TypeError:
+                return NotImplemented
+        return self._compare_term(other, operator.lt)
+
+    def __hash__(self):
+        return hash((self.exp, self.mult))
+
+class Ordinal(Basic):
+    """
+    Represents ordinals in Cantor normal form.
+
+    Internally, this class is just a list of instances of OmegaPower
+    Examples
+    ========
+    >>> from sympy.sets import Ordinal, ord0, OmegaPower
+    >>> from sympy.sets.ordinals import omega
+    >>> w = omega
+    >>> w.is_limit_ordinal
+    True
+    >>> Ordinal(OmegaPower(w + 1 ,1), OmegaPower(3, 2)) #doctest: +SKIP
+    w**(w + 1) + w**3*2
+    >>> 3 + w  #doctest: +SKIP
+    w
+    >>> (w + 1) * w #doctest: +SKIP
+    w**2
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Ordinal_arithmetic
+    """
+    def __new__(cls, *terms):
+        obj = super(Ordinal, cls).__new__(cls, *terms)
+        powers = [i.exp for i in obj.args]
+        if not all(earlier >= later for earlier, later in zip(powers, powers[1:])):
+            raise ValueError("powers must be in decreasing order")
+        return obj
+
+    @property
+    def is_successor_ordinal(self):
+        return self.args[-1].exp == ord0
```

Comment:
I have often found that using names for `self.args[some_index]` if can be assigned to some `self.attribute_name` via `@property` decorator is always better, since otherwise if in future API is change it create quite some problems.

Also `self.property_name` is more clear indication of what `property_name` is supposed to mean opposed to `.args[-1]`.

Reply:
In this case `args` is a tuple of varying length and it is the last item that should be accessed.

But I see that the case of empty `args` should be taken care of.